docs/manual/tsort.md

```diff
@@ -2,160 +2,69 @@
-all of the other dependencies are simply not needed in an upgrade context
-to perform package ordering. Please note that all of the known to cause
-loop dependencies listed above are, for now, explicitly ignored when
-determining package install ordering.
-
-## Summary
-
-So what does this all mean? Basically not much, unless you find yourself
-trying to specify dependencies amongst a set of packages correctly and
-happen to create a dependency loop.
-
-And, before you start adding the new-fangled syntax to packages, please
-remember that rpm will almost certainly be auto-generating fine-grained
-dependencies for %post et al scriptlets pretty soon. Truly, rpm needs to
-make packaging easier, not provide Yet More Complicated Syntax in spec files.
-
-With thanks to Ken Estes for doing the implementation in bash2 that makes
-it possible to auto-generate scriptlet dependencies, blame me for the long,
-slow deployment.
+# Package ordering
+
+Rpm orders transactions solely based on package dependency information:
+dependencies are installed before the dependent, and erased after it.
+If the install/erase order is not right, then either the dependency
+information is incomplete, incorrect or there are unresolvable dependency
+loops.
+
+The order is calculated by matching dependencies on `Provides` from the
+packages in the transaction. If package A `Requires` B, then a provider
+of B will be installed before A, and erased after it.
+In case of multiple providers, heuristics are used to pick the best match.
+Weak dependencies (`Recommends`, `Suggests`, `Supplements`, `Enhances`)
+and ordering hints (`OrderWithRequires`) are similarly considered, if the
+provider is present in the transaction.
+Note that `Supplements` and `Enhances` are reverse dependencies and are
+ordered thus.`Obsoletes` and `Conflicts` dependencies are not used for
+ordering.
+
+A dependency loop is generally a packaging bug as no correct ordering is
+possible to achieve in that situation. They are best solved at the source,
+by splitting packages to sub-packages to eliminate the loop.
+Sometimes this is not practical though, and also some dependencies are
+not relevant for ordering at all. Rpm supports various
+[dependency qualifiers](spec.md#Requires) to supply
+context that rpm uses for loop resolution.
+
+Qualifiers work in two ways, depending on the context and the qualifier:
+they can either *resolve* the loop, or *favor* loop cutting.
+
+Normally, `Requires: A` and `Requires(pre): A` result in exactly the same
+install order but when a loop is detected, the latter is favored during
+installation. During erasure, the latter is ignored.
+In simple cases this is sufficient to nudge the loop resolution in the
+desired direction, but there is no guarantee. Also note that favoring
+is only used on hard `Requires` dependencies.
+
+`Requires(preun): A` declares the dependency is only needed during erasure,
+so it can resolve an install-time loop. Unordering qualifiers cause the
```

Comment:
May be "non-ordering" is a better term here? May be a native speaker can chime in on that.

Reply:
I chose unorder intentionally as those qualifiers *unorder* the dependency. Names are hard, native speakers input welcome of course.

Reply:
The fundamental problem is that rpm needs to build a DAG (partial order) out of the dependencies.  But to create a DAG, there can't be circular dependencies (i.e., loops).  To create a DAG, rpm breaks loops by removing edges.  Because removing an edge means that a package will be installed (or removed) before a package that it depends on is installed, this is fragile.  rpm provides a mechanisms to influence how it should break these dependency loops.

Given the above: instead of using the term unordering qualifier, perhaps dependency refinement would be better.

Reply:
We use the term qualifier for the "stuff between the parentheses" in other places, so I think it makes sense to stick to the terminology even if less than perfect.